Modules/_heapqmodule.c

```diff
@@ -538,6 +538,460 @@ _heapq__heapify_max(PyObject *module, PyObject *heap)
+        PyObject *tmpargs = PyTuple_New(0);
+        if (tmpargs == NULL)
+            return NULL;
+        if (!PyArg_ParseTupleAndKeywords(tmpargs, kwds, "|Op:merge",
+                                         kwlist, &keyfunc, &reverse)) {
+            Py_DECREF(tmpargs);
+            return NULL;
+        }
+        Py_DECREF(tmpargs);
+    }
+
+    if (keyfunc == Py_None)
+        keyfunc = NULL;
+    Py_XINCREF(keyfunc);
+
+    if (keyfunc != NULL && !PyCallable_Check(keyfunc)) {
+        PyErr_SetString(PyExc_TypeError, "Key must be callable or None.");
+        goto error;
+    }
+
+    assert(PyTuple_CheckExact(args));
+    num_iters = PyTuple_GET_SIZE(args);
+
+    if (num_iters == 0) {
+        mo = (mergeobject *)type->tp_alloc(type, 0);
+        if (mo == NULL)
+            return NULL;
+        mo->iterators = NULL;
+        mo->tree = NULL;
+        mo->keyfunc = NULL;
+        mo->reverse = 0;
+        mo->started = 1;
+        return (PyObject *) mo;
+    }
+
+    iterators = PyList_New(num_iters);
+    if (iterators == NULL)
+        goto error;
+
+    /* Store the tree as a list with n-1 internal nodes and n leaves.
+       If a key is being used, store keys next to original items. */
+    tree = PyList_New(keyfunc? num_iters * 4 - 2: num_iters * 2 - 1);
+    if (tree == NULL)
+        goto error;
+
+    /* To ensure stability, we need to order the iterators so that the
+     * iterator derived from the first iterable passed is exactly the iterator
+     * that supplies the leftmost leaf node in the tree. If we shift the array
+     * accordingly, then we can always supply tree[i] with elements from
+     * iterators[i-(n-1)].
+     */
+    shift = get_iter_shift(num_iters);
+    assert(shift >= 0);
+
+    /* Now make the shifted list of iterators. */
+    for (i = 0; i < num_iters; i++) {
+        PyObject *iterable = PyTuple_GET_ITEM(args, (i + shift) % num_iters);
+        PyObject *iterator = PyObject_GetIter(iterable);
+        if (iterator == NULL)
+            goto error;
+        PyList_SET_ITEM(iterators, i, iterator);
+    }
+
+    mo = (mergeobject *)type->tp_alloc(type, 0);
+    if (mo == NULL)
+        goto error;
+    mo->iterators = iterators;
+    mo->tree = tree;
+    mo->keyfunc = keyfunc;
+    mo->reverse = reverse;
+    mo->started = 0;
+
+    return (PyObject *) mo;
+
+error:
+    Py_XDECREF(keyfunc);
+    Py_XDECREF(iterators);
+    Py_XDECREF(tree);
+    return NULL;
+}
+
+static void
+merge_dealloc(mergeobject *mo)
+{
+    PyObject_GC_UnTrack(mo);
+    Py_XDECREF(mo->tree);
+    Py_XDECREF(mo->iterators);
+    Py_XDECREF(mo->keyfunc);
+    Py_TYPE(mo)->tp_free(mo);
+}
+
+static PyObject *
+merge_sizeof(mergeobject *mo, void *unused)
+{
+    Py_ssize_t res = _PyObject_SIZE(Py_TYPE(mo));
+
+    if (mo->iterators)
+        res += PyList_GET_SIZE(mo->iterators) * sizeof(PyObject *);
+    if (mo->tree)
+        res += PyList_GET_SIZE(mo->tree) * sizeof(PyObject *);
```

Comment:
We are storing also the size of the `keyfunc` object, no?

Reply:
I was under the impression that `__sizeof__` should not recursively include the sizes of its members. I was only including the sizes of `iterators` and `tree` because they aren't externally accessible and I was treating them as private arrays. Is that correct?

Reply:
> I was under the impression that __sizeof__ should not recursively include the sizes of its members.

It would not be doing it recursively, it just uses a pointer to a PyObject. But you are right in the sense that it should not be accounted for because the pointers are already accounted for by `object.__sizeof__`  when it checks `self->ob_type->tp_basicsize`.